### ISSUE TITLE
add option (timeout) to switch from drain to drain with disable-eviction

### ISSUE THREAD

**Issue by damoon** [weaveworks/kured#303]
Details for the disable-eviction option
https://github.com/kubernetes/kubernetes/issues/83307
https://github.com/kubernetes/kubernetes/pull/85571

LD/DR
Allows to drain pod from a node while ignoring PodDisruptionBudgets.
Included in k8s 1.18.

Use case:
We run Kubernetes for CI. A lot of test environments contain broken containers. It is almost impossible to drain pods from a node respecting the pdb. For each node we need manual intervention and delete a few "bad" pods by hand.
I would like to have the option to say: After e.g. 1 hour of draining without success, drain the pods ignoring the PodDisruptionBudgets.
This would allow me to reboot nodes without supervision and free up a lot of time.

**Comment by evrardjp** [comment]
Let's keep this opened

**Comment by evrardjp** [comment]
It's a good feature, I think we probably want to expose this. Should we target this for 1.8.0? Will you tackle it @damoon ? That would be awesome.

**Comment by atighineanu** [comment]
I might have a glance at this feature. Testing/reproducing such an environment, however, might take longer than writing the feature itself :)

**Comment by ffilippopoulos** [comment]
The ability to drain nodes without respecting PDBs is an important feature for us as well. Currently the only way to really use kured in dev environments is by using `--force-reboot` flag, which is not ideal. Ideally, after the drain timeout is reached, we could try to drain with disable-eviction and reboot without releasing the lock at all.

**Comment by ckotzbauer** [comment]
We would be happy to receive a PR for this. It seems to be a useful feature 👍

**Comment by ffilippopoulos** [comment]
reading a bit more regarding this, I think that kube's graceful shutdown: https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown should cover us in case we use `--force-reboot` and delete any pods stuck in the node before attempting a reboot. We'll try this for a while in our lower envs and if we see any issue I am happy to PR an optional disable evictions flag for the drainer.

**Comment by fa-at-pulsit** [comment]
Maybe it is better instead to implement the following logic, like in this [article](https://medium.com/@tamber/solution-avoid-kubernetes-openshift-node-drain-failure-due-to-active-poddisruptionbudget-df68efed2c4f):

1. collect all PDBs in the Cluster
```
kubectl get pdb -A -o json | jq -r '.items[] | select(.spec.minAvailable >= 1)  | [.metadata.name , .metadata.namespace, .spec.minAvailable] | @tsv' >> pdb_namespace_originalMinimalAvailable_list.txt
```
2. patch the existing PDBs to have minAvailable=0
```
while read p; do \
PDB=$(echo $p | awk '{print $1}') ; \
echo PDB: ${PDB} ; \
NS=$(echo $p | awk '{print $2}') ; \
echo NS: ${NS} ; \
kubectl patch pdb $PDB -n $NS --type=merge -p '{"spec":{"minAvailable":0}}' ; \
done <  pdb_namespace_originalMinimalAvailable_list.txt
```
3. drain & reboot
4. restore  original minAvailable setting
```
while read p; do \
PDB=$(echo $p | awk '{print $1}') ; \
echo PDB: ${PDB} ; \
NS=$(echo $p | awk '{print $2}') ; \
echo NS: ${NS} ; \
Original_Min_Available=$(echo $p | awk '{print $3}') ; \
echo  Original_Min_Available: ${Original_Min_Available} ; \
command=`echo "kubectl patch pdb ${PDB} -n ${NS} --type=merge -p 
'{\"spec\":{\"minAvailable\":${Original_Min_Available}}}'"` ;\
echo ${command} ;\
eval ${command} ;\
done < pdb_namespace_originalMinimalAvailable_list.txt
```

**Comment by ckotzbauer** [comment]
@fa-at-pulsit I strongly disagree to implement such a logic in kured. This would mean that kured would change the settings of all PDBs in a way, that they loose there effect. This is the same behaviour as just deleting them (and restore them afterwards). This raises the question of why the cluster operator introduced PDBs in the first place then. Kured should not touch PDBs at all.

I'm :+1: to implement a timeout-solution as @damoon initially suggested, but all applications must have the possibility of a graceful shutdown.

**Comment by fa-at-pulsit** [comment]
@ckotzbauer, yes, you are right.... from this side, timeout-solution will be better!
